File: GoCoder/answer/15_3Sum.go
package answer

import (
	"fmt"
	"sort"
)

func ThreeSum(nums []int) [][]int {
	result := [][]int{}
	sort.Ints(nums)
	if len(nums) < 3{
		return [][]int{}
	}
	for first, _ := range nums[:len(nums)-2]{
		if first>0 && nums[first] == nums[first-1]{ //选择第一个未重复的元素
			continue
		}
		third := len(nums) - 1
		second:=first+1
		for second != third{
			sumNow := nums[first] + nums[second] + nums[third]
			if sumNow == 0{
				result = append(result, []int{nums[first], nums[second], nums[third]})
			} else if sumNow > 0{
				for second != third{
					third--
					if nums[third] != nums[third+1]{
						break
					}
				}
				continue
			}
			for second != third{ // sumNow < 0 和sumNom == 0时都右移second
				second++
				if nums[second] != nums[second-1]{
					break
				}
			}
		}
	}
	fmt.Println(result)
	return result
}

File: PyCoder/answer/15_3Sum.py
from typing import List


class Solution15:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums = sorted(nums)
        print(nums)
        result = []
        if len(nums) < 3:
            return result
        for first in range(len(nums)-2):
            if first > 0 and nums[first] == nums[first-1]:
                continue
            second = first + 1
            third = len(nums) - 1
            while second < third:
                sumNow = nums[first] + nums[second] + nums[third]
                if sumNow == 0:
                    result.append([nums[first], nums[second], nums[third]])
                elif sumNow < 0:
                    while second < third and nums[second] == nums[second+1]:
                        second += 1
                    second += 1
                    continue
                while second < third and nums[third] == nums[third-1]:
                    third -= 1
                third -= 1
        return result


if __name__ == "__main__":
    s = Solution15()
    print(s.threeSum([1, 0, -1, 2, -1]))
